dvc/remote/local/__init__.py

```diff
@@ -581,3 +581,76 @@ def _get_unpacked_dir_names(self, checksums):
             if self.is_dir_checksum(c):
                 unpacked.add(c + self.UNPACKED_DIR_SUFFIX)
         return unpacked
+
+    def _needs_checkout(self, path_info, checksum):
+        # NOTE: In case if path_info is already cached, and cache type is
+        # 'copy' we would like to avoid relinking.
+        # `symlink/hardlink/reflink` is cheap, but if we need `copy` links,
+        # then to save on copying we should only relink if we already have
+        # `hardlink` or a `symlink`, we don't care about reflinks, because
+        # they are indistinguishable from copy anyway.
+
+        if (
+            # self._is_cache_copy(path_info)
+            not self.changed(path_info, {self.PARAM_CHECKSUM: checksum})
+            and self._is_same_link_as_cache(path_info)
+            # and not System.is_hardlink(path_info)
+            # and not System.is_symlink(path_info)
+        ):
+            msg = "File '{}' didn't change"
+            logger.debug(msg.format(str(path_info)))
+            return False
+        return True
+
+    def _is_same_link_as_cache(self, path_info):
+        is_hardlink = System.is_hardlink(path_info)
+        is_symlink = System.is_symlink(path_info)
+        is_copy_or_reflink = not is_hardlink and not is_symlink
+
+        if self.cache_types[0] == "reflink":
+            workspace_file = path_info.with_name("." + uuid())
+            test_cache_file = self.path_info / ".cache_type_test_file"
+            if not self.exists(test_cache_file):
+                with open(fspath_py35(test_cache_file), "wb") as fobj:
+                    fobj.write(bytes(1))
+            try:
+                self.link(test_cache_file, workspace_file)
+            finally:
+                self.remove(workspace_file)
+                self.remove(test_cache_file)
+
+        if self.cache_types[0] == "symlink":
+            return is_symlink
+        if self.cache_types[0] == "hardlink":
+            return is_hardlink
+
+        if self.protected:
+            self.protect(path_info)
+        else:
+            # NOTE: we can unprotect, because `hardlink/symlink` check
```

Comment:
We also should protect/unprotect symlinks/hardlinks in this case, right? Since we are detecting if something is already a *link, but not detecting if it is protected or not.

Reply:
What in case of [sym, hard] link and unprotected repo? It is possible to have hardlink cache and not have it protected. If we were to unprotect if repo is not protected, we might as well have copy link type.

Reply:
But "copy" will lead to data duplication, which is not something that happens in unprotected repo with *links.

Reply:
W8, by data duplication we mean copying file from cache, right?

It seems to me that in case of `*link` there is [data duplication on unprotect](https://github.com/iterative/dvc/blob/41351a2df2f36ed04846d5e3d976d6d99b71a151/dvc/remote/local/__init__.py#L475).

Reply:
@pared on `dvc unprotect` - yes, since it copies your file. But when you have `=protected False` in your config with `hardlink` or `symlink` cache type, we don't use copies instead right? That is why in this case we should protect or unprotect(no copying) the link to match how it would look like if the user just started his repo with `protected False` and `hardlink/symlink`.

Reply:
Thats true, I will fix that.